### DISCUSSION TITLE
Merging bag/rdd data

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#274] [Design & Development]
I ran across something that struck me as useful, but is something for which I don't see a clear solution.   I'm hoping one of you know if what I'm searching for exists.

The generic problem for MsPASS is this: there are a large class of algorithms for which the output is readily reduced to a few numbers (sometimes as small as a single number).  For now the only way I see to handle that kind of algorithm in a map operation is to have the function post the number(s) to the Metadata container for the object before return the data modified by updating Metadata.   That can work fine, but strikes me as a potential inefficiency.   When dask or Spark have to move such output between nodes we will invoke the overhead of pickling and then unpickling the output.  

What I am searching for is the equivalent of a one-to-one join of two bags/RDDs.   i.e. in this model one would probably need to require that every data object produced some kind of output even if it was an error code.  Then one could require a function that would take the same positional value in two bags and merge them.   To be more specific, suppose we had a function that would take TimeSeries input and compute a single number (e.g. an automated phase picker that would return an arrival time).   We could run the function as is with a map to produce a new bag/rdd that is only a list of phase pick times.   A merge function would be easy to write that would merge the TimeSeries/Seismogram input container with corresponding times returned in the second container from the picker.   That should be feasible, I think, since a bag/rdd is just a list of items. 

I am not able to find such a capability in either dask or spark.   The documentation can be pretty cryptic, so it is quite possible have just not found the correct method.   Alternately, I'm asking for something that isn't feasible.   I can see this would not be possible as described if the output of a map operator is not necessarily in the same order as the input.   

If what I seem to find to be finding is true, which is that the merge operation I described above can't be done, then it seems we have only two options:
1.  Recognize and document that the only alternative for such function is to post the answer(s) to the data object's metadata container.  
2. Use MongoDB and do a transaction for each computed output.   e.g. a pick function might generate and document in an "arrival" collections.

Both of the above options have potential performance issues so I would like to know if any of you know if there is a merge procedure we could use with spark and/or dask.

**Comment by wangyinz** [comment]
With what we have implemented so far, it is indeed an inefficiency. However, the inefficiency is caused by our implementation instead of the use of Dask/Spark. Both packages have the API to achieve what you described above - join with a iterable object by a specified key. Although the API is very different, both Dask and Spark call this operation `join` ([ref](https://docs.dask.org/en/latest/generated/dask.bag.Bag.join.html#dask.bag.Bag.join), [ref](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.join.html)). We will need to wrap around the call to create a temporary key and merge the new attribute into `Metadata` to achieve this functionality in a user-friendly way. 

This is certainly something we could implement, but it should also be considered an improvement rather than a core functionality, because user can always use the native API of Dask/Spark to achieve what they need. If not, they can just do the inefficient Metadata update. I actually think that it may not be too inefficient in most cases.
